File: problems/0022-generate-parentheses/solution.py
"""
Problem Number: 22
Problem Name: Generate Parentheses
Difficulty: Medium
Topics: String, Dynamic Programming, Backtracking
Companies: Facebook, Microsoft, Amazon
"""

from typing import List

class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        """
        Approach: Backtracking
        Time Complexity: O(4^n / sqrt(n))
        Space Complexity: O(n)
        """
        res = []
        
        def backtrack(s: str, open_count: int, close_count: int):
            if len(s) == 2 * n:
                res.append(s)
                return
            
            if open_count < n:
                backtrack(s + "(", open_count + 1, close_count)
            
            if close_count < open_count:
                backtrack(s + ")", open_count, close_count + 1)
                
        backtrack("", 0, 0)
        return res

# Test cases
if __name__ == "__main__":
    solution = Solution()
    print(solution.generateParenthesis(3)) # ["((()))","(()())","(())()","()(())","()()()"]
    print(solution.generateParenthesis(1)) # ["()"]

File: problems/0022-generate-parentheses/solution.js
/**
 * @problemNumber: 22
 * @problemName: Generate Parentheses
 * @difficulty: Medium
 * @topics: String, Dynamic Programming, Backtracking
 * @companies: Facebook, Microsoft, Amazon
 */

/**
 * Approach: Backtracking
 * Time Complexity: O(4^n / sqrt(n))
 * Space Complexity: O(n)
 *
 * @param {number} n
 * @return {string[]}
 */
function generateParenthesis(n) {
    const result = [];

    function backtrack(currentStr, openCount, closeCount) {
        if (currentStr.length === 2 * n) {
            result.push(currentStr);
            return;
        }

        if (openCount < n) {
            backtrack(currentStr + "(", openCount + 1, closeCount);
        }

        if (closeCount < openCount) {
            backtrack(currentStr + ")", openCount, closeCount + 1);
        }
    }

    backtrack("", 0, 0);
    return result;
}

// Test cases
console.log(generateParenthesis(3)); // ["((()))","(()())","(())()","()(())","()()()"]
console.log(generateParenthesis(1)); // ["()"]

module.exports = { generateParenthesis };
